[rpc] getblockchaininfo: add size_on_disk, prune_target_size

esotericnonsense opened the pull request.

esotericnonsense commented:
If #11359 gets in it might be nice to have the prune hwm in there too, but I didn't want to add the dependency.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1137,6 +1137,8 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
             "  \"chainwork\": \"xxxx\"     (string) total amount of work in active chain, in hexadecimal\n"
             "  \"pruned\": xx,             (boolean) if the blocks are subject to pruning\n"
             "  \"pruneheight\": xxxxxx,    (numeric) lowest-height complete block stored\n"
+            "  \"disk_size\": xxxxxx,    (numeric) the estimated size of the block and undo files on disk\n"
```
nit: perhaps `disk_usage` or `size_on_disk`?

Also consider placing this above `pruned` so all of the pruning-related fields are grouped.

promag replied:
Also align description.

promag replied:
IMO we should test the response result. In this case, at least, check that the key exists and value is a number. cc @jnewbery

promag replied:
Ah, @jnewbery already recommended a test! 👍

jnewbery reviewed (commented):
concept ACK. One nit inline.

Consider adding an explicit test for the `getblockchaininfo` RPC method to `test/functional/blockchain.py`.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1180,6 +1182,7 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
     obj.push_back(Pair("verificationprogress",  GuessVerificationProgress(Params().TxData(), chainActive.Tip())));
     obj.push_back(Pair("chainwork",             chainActive.Tip()->nChainWork.GetHex()));
     obj.push_back(Pair("pruned",                fPruneMode));
+    obj.push_back(Pair("disk_size",             CalculateCurrentUsage()));
```
Missing lock of `cs_LastBlockFile`.

promag commented:
Test for `getblockchaininfo` added in #11370.

esotericnonsense commented:
A few mishaps with the comment message there, sorry.

I have merged this with #11366 and closed #11366.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1135,8 +1135,10 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
             "  \"chainwork\": \"xxxx\"     (string) total amount of work in active chain, in hexadecimal\n"
+            "  \"size_on_disk\": xxxxxx,   (numeric) the estimated size of the block and undo files on disk\n"
             "  \"pruned\": xx,             (boolean) if the blocks are subject to pruning\n"
             "  \"pruneheight\": xxxxxx,    (numeric) lowest-height complete block stored\n"
```
The help text for `pruneheight` should also include "(only present if pruning is enabled)"

esotericnonsense replied:
4a61af9

jnewbery reviewed (commented):
one more documentation nit. Otherwise looks good to me

esotericnonsense committed:
[rpc] getblockchaininfo: add size_on_disk, prune_target_size, automatic_pruning

Fix pruneheight help text.
Move fPruneMode block to match output ordering with help text.
Add functional tests for new fields in getblockchaininfo.

sipa commented on src/validation.h:
```diff
@@ -156,6 +156,7 @@ struct BlockHasher
 
 extern CScript COINBASE_FLAGS;
 extern CCriticalSection cs_main;
+extern CCriticalSection cs_LastBlockFile;
```
I'd rather not expose more variables that are private to validation to the outside world.

Perhaps you can instead create a wrapper around `CalculateCurrentUsage`, which just grabs `cs_LastBlockFile` and calls the internal one, and expose that? That way the RPC code wouldn't need to know about the existence of that lock eve.

promag replied:
Why not just lock `cs_LastBlockFile` in `CalculateCurrentUsage`? It's a recursive mutex and IMO should be locked where it's needed (BTW the same for remaining locks...).

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1179,7 +1181,18 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
+    obj.push_back(Pair("size_on_disk",          CalculateCurrentUsage()));
     obj.push_back(Pair("pruned",                fPruneMode));
+    if (fPruneMode) {
+        CBlockIndex *block = chainActive.Tip();
```
`CBlockIndex* block = ...`;

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1168,7 +1170,7 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
         );
 
-    LOCK(cs_main);
+    LOCK2(cs_main, cs_LastBlockFile); // cs_LastBlockFile for CalculateCurrentUsage()
```
As @sipa already commented, `CalculateCurrentUsage()` should be responsible for concurrency locking (not the calling code part).

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1135,8 +1135,10 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
             "  \"pruned\": xx,             (boolean) if the blocks are subject to pruning\n"
-            "  \"pruneheight\": xxxxxx,    (numeric) lowest-height complete block stored\n"
+            "  \"pruneheight\": xxxxxx,    (numeric) lowest-height complete block stored (only present if pruning is enabled)\n"
+            "  \"prune_target_size\": xxxxxx,  (numeric) the target size used by pruning (only present if pruning is enabled)\n"
```
Indentation seems wrong (others need to space-up).

esotericnonsense replied:
I followed the example of 'verificationprogress'; can indent the others further if you think that makes sense.

jonasschnelli commented:
Concept ACK

esotericnonsense commented:
Rebased on master, moved the lock into CalculateCurrentUsage, `CBlockIndex *block` => `CBlockIndex* block`

sipa commented:
utACK 071879a6263aa286e166f9cf07f409e6df35bc02

jnewbery commented:
I've just tested this with `-prune=1` and I get the bizarre result:

```
→ bitcoin-cli getblockchaininfo
{
...
  "prune_target_size": 18446744073709551615,
```

That's because of this section in `init.cpp`:

```
    nPruneTarget = (uint64_t) nPruneArg * 1024 * 1024;
    if (nPruneArg == 1) {  // manual pruning: -prune=1
        LogPrintf("Block pruning enabled.  Use RPC call pruneblockchain(height) to manually prune block and undo files.\n");
        nPruneTarget = std::numeric_limits<uint64_t>::max();
        fPruneMode = true;
...
```

Returning the max int for `prune_target_size` is meaningless and confusing. I recommend you don't return `prune_target_size` if bitcoind is configured for manual pruning (ie if `-prune=1`) (and perhaps also add *another* field `pruning_mode` that returns whether pruning is automatic or manual).

esotericnonsense commented:
Hmmm. Yes, that isn't desirable behaviour.

Options I can see:
a) nPruneArg is exposed from init.cpp, or an fPruneManualMode (in which case it could be explicitly checked for in FindFilesToPrune)
or
b) the RPC checks for int_max to determine whether automatic pruning is disabled

b) fixes the RPC without having to touch pruning code but otherwise seems a bit messy to me.

MeshCollider commented:
Or instead of b) just see if gArgs.GetArg("-prune", 0) is 1?

jnewbery commented:
I think just testing the value of `gArgs.GetArg("-prune", 0)` is appropriate for this pr. No need to touch the init or valuation code.

esotericnonsense commented:
```
$ grep "prune" ~/.bitcoin/bitcoin.conf
prune=550
$ ./bitcoin-cli getblockchaininfo | grep "prune"
  "pruned": true,
  "pruneheight": 0,
  "pruneauto": true,
  "prune_target_size": 576716800,

$ grep "prune" ~/.bitcoin/bitcoin.conf
prune=1
$ ./bitcoin-cli getblockchaininfo | grep "prune"
  "pruned": true,
  "pruneheight": 0,
  "pruneauto": false,

$ grep "prune" ~/.bitcoin/bitcoin.conf
prune=0
$ ./bitcoin-cli getblockchaininfo | grep "prune"
  "pruned": false,
```

promag commented:
Please rebase with #11370.

esotericnonsense commented:
Rebased on master and added tests for the relevant fields in test/functional/blockchain.py.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1179,7 +1182,23 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
+
+        obj.push_back(Pair("pruneheight",        block->nHeight));
+
+        bool fPruneAuto = (gArgs.GetArg("-prune", 0) > 1);
```
current code style is snake_case for variables (https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#developer-notes). `prune_auto` or `auto_prune` should be fine.

jnewbery replied:
nit: Since `1` is a magic number meaning manual pruning, I think it's clearer to use equality like in init.cpp, rather than greater-than:

```
        bool fPruneAuto = (gArgs.GetArg("-prune", 0) != 1);
```

esotericnonsense replied:
I was torn there, because that actually gives True in the case prune=0, but that case can never happen because you'd skip the if (fPruneMode) block. I'll change it.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1179,7 +1182,23 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
+        obj.push_back(Pair("pruneheight",        block->nHeight));
+
+        bool fPruneAuto = (gArgs.GetArg("-prune", 0) > 1);
+        obj.push_back(Pair("pruneauto",          fPruneAuto));
```
Please also use snake_case for RPC fields. I think `automatic_pruning` is appropriate here.

jnewbery reviewed (commented):
Implementation looks good. I've slightly tested and it works well.

A few style nits inline.

esotericnonsense commented on test/functional/blockchain.py:
```diff
@@ -58,20 +58,42 @@ def _test_getblockchaininfo(self):
+        assert_equal(sorted(res.keys()), sorted(['pruneheight', 'pruneauto'] + keys))
+
+        # size_on_disk should be >= 0
+        assert res['size_on_disk'] >= 0
```
As below I've dropped the 'equal to' because I don't see why it should ever be 0 unless there's some caching going on.

esotericnonsense commented:
Changes made and rebased on master (second push, first was a mistake).

jnewbery commented:
Tested ACK 2b73ece6e34961a6745114b6932945692ed987e7

TheBlueMatt reviewed (commented):
utACK 2b73ece6e34961a6745114b6932945692ed987e7

MarcoFalke commented:
Needs rebase™

esotericnonsense commented on src/validation.cpp:
```diff
@@ -3230,8 +3230,10 @@ bool TestBlockValidity(CValidationState& state, const CChainParams& chainparams,
-static uint64_t CalculateCurrentUsage()
+uint64_t CalculateCurrentUsage()
 {
+    LOCK(cs_LastBlockFile);
```
Done in b7dfc6c4b89b62f9bb79ea009ee103a6299ac005.

esotericnonsense commented:
®ebased

MarcoFalke commented:
utA©K b7dfc6c4b89b62f9bb79ea009ee103a6299ac005

TheBlueMatt commented:
re-utACK b7dfc6c4b89b62f9bb79ea009ee103a6299ac005. In the future, can you avoid rebasing when squashing unless you need to? Avoiding rebasing onto latest master makes it easier for reviewers to identify what changed between reviews.

esotericnonsense commented:
Yes, I can do that in the future. I hadn't considered that.
Do you prefer for each change to be made in an individual commit (with a squash before the PR is implemented)?

MarcoFalke commented:
It is fine to just amend the existing commits with fix-ups, as reviewers
have the original commits fetched locally. So it is easy to compare the two
versions.

However, with a rebase on master in between, the reviewer needs to
painfully repeat the same rebase in some way.
